docs/frontend-protocol.rst

```diff
@@ -0,0 +1,181 @@
+**Be careful**: a malicous "other end" could cause all sorts of shenanigans.
+
+You are only limited to streaming endpoints (i.e. no UDP) but they do not have to be TCP.
+For example, one side could use ``unix:/tmp/socket`` to open a listener (or connection) on Unix-domain socket.
+With `txtorcon <https://meejah.ca/projects/txtorcon>`_ one could have ``onion:...`` Tor endpoints on either end.
+
+Once the listener is established, we'll issue a ``kind: listening`` output.
+
+
+Input: ``kind: remote-to-local``
+--------------------------------
+
+This will cause ``fowld`` to request a listener on the *other* side.
+There is symmetry here: the same thing could be accomplished by that other side instead issuing a ``kind: local`` request.
+
+Keys allowed in this message:
+
+- ``listen-endpoint`` (required): a Twisted server-style endpoint string specifying where to listen (on the other end).
+- ``connect-endpoint`` (required): a Twisted client-style endpoint string specifying what to connect to (on this side) for each connection that happens on the other side.
+
+To directly mirror the example from the ``local`` command:
+
+.. code-block:: json
+
+    {
+        "kind": "remote-to-local",
+        "listen-endpoint": "tcp:8000:interface=localhost",
+        "connect-endpoint": "tcp:localhost:80"
+    }
+
+This will be a mirror-image of the other example.
+That is, we'll cause the far end to start listening on its TCP port ``8000`` on interface ``localhost``.
+Any connection to that will open a near-side connection to port 80 via TCP.
+
+The far-side ``fowld`` will issue a ``kind: listening`` message when it has started listening.
+
+
+Output: ``kind: listening``
+-----------------------------
+
+This message is issued by ``fowld`` when it has opened a listening socket on that side.
+
+So, if a ``kind: local`` had initiated the listening, this message would appear on that same side.
+If instead it was a ``kind: remote-to-local`` then it would appear on the far side.
+
+An example message:
+
+.. code-block:: json
+
+    {
+        "kind": "listening",
+        "endpoint": "tcp:8080:interface=localhost",
+        "connect-endpoint": "tcp:80"
+    }
+
+Guidance for UX: the user should be made aware their machine is listening on a particular port / interface.
+
+
+Output: ``kind: error``
+--------------------------
+
+Some sort of error has happened.
+
+This message MUST have a ``message`` key containing a freeform error message.
+It MAY have additional fields depending on the kind of error (XXX good idea? Hard to produce strict parser...)
+
+An example message:
+
+.. code-block:: json
+
+    {
+        "kind": "error",
+        "message": "Unknown control command: foo"
+    }
+
+Guidance for UX: most errors are probably interesting to the user.
+
+
+Output: ``kind: connected``
+-----------------------------
+
+The ``fowld`` process has successfully connected to the Mailbox Server.
```

Comment:
At least right now this message gets output only after the other side connects and does not really refer to the connection to the Mailbox Server I think.